src/ledger/LedgerStateOfferSQL.cpp

```diff
@@ -0,0 +1,344 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "crypto/KeyUtils.h"
+#include "crypto/SecretKey.h"
+#include "database/Database.h"
+#include "ledger/LedgerEntryReference.h"
+#include "ledger/LedgerState.h"
+#include "util/types.h"
+
+namespace stellar
+{
+using xdr::operator==;
+using xdr::operator<;
+using xdr::operator>;
+
+void
+processAsset(Asset& asset, AssetType assetType, std::string const& issuerStr,
+             soci::indicator const& issuerIndicator,
+             std::string const& assetCode,
+             soci::indicator const& assetCodeIndicator)
+{
+    asset.type(assetType);
+    if (assetType != ASSET_TYPE_NATIVE)
+    {
+        if ((assetCodeIndicator != soci::i_ok) ||
+            (issuerIndicator != soci::i_ok))
+        {
+            throw std::runtime_error("bad database state");
+        }
+
+        if (assetType == ASSET_TYPE_CREDIT_ALPHANUM12)
+        {
+            asset.alphaNum12().issuer =
+                KeyUtils::fromStrKey<PublicKey>(issuerStr);
+            strToAssetCode(asset.alphaNum12().assetCode, assetCode);
+        }
+        else if (assetType == ASSET_TYPE_CREDIT_ALPHANUM4)
+        {
+            asset.alphaNum4().issuer =
+                KeyUtils::fromStrKey<PublicKey>(issuerStr);
+            strToAssetCode(asset.alphaNum4().assetCode, assetCode);
+        }
+        else
+        {
+            throw std::runtime_error("bad database state");
+        }
+    }
+}
+
+LedgerState::StateEntry
+LedgerState::loadOfferFromDatabase(LedgerKey const& key)
+{
+    auto& db = mRoot->getDatabase();
+
+    uint64_t offerID = key.offer().offerID;
+    std::string actIDStrKey = KeyUtils::toStrKey(key.offer().sellerID);
+
+    std::string sql = "SELECT sellerid, offerid, "
```

Comment:
undo the duplication of this sql string like it was before as all calls to `loadOffersFromDatabase` must exactly match when it comes to columns being returned.